### PR TITLE
Update kmeans neighborhood notebook take 2

### PR DESCRIPTION
**Pull request by cliu72**
**What is the purpose of this PR?**

Update and clean-up the k-means neighborhood notebook.

**How did you implement your changes**

- Fixed a bug where the wrong cell phenotypes were being counted due to fovs missing certain cell phenotypes.
-  The old notebook required the cell table to include both a column that specified a numeric id for each cell phenotype, and a column that had the cell phenotypes. I simplified this to only require the column with the cell phenotypes. Specifically, in create_neighborhood_matrix, removed cluster_id_col, which was redundant with cluster_name_col.
- Set default of self_neighbor to False, so a cell is not counting itself as its neighbor.
- Set cluster numbers to start at 1. Previously was starting at 0, which is confusing because in cell overlays, 0 usually corresponds to empty slide.
- Add inertia as a metric.
- Subsample cells for silhouette score calculation.
- Remove cells that have no neighbors before clustering.
- Scaled (z-scored) the heatmaps, which makes them more interpretable.
- Added heatmap that shows the composition of each neighborhood (basically the average feature vector for each cluster).
- Fixed overlay generation function so correct legend is shown.
- Added subheadings and more descriptive text to notebook.

**Remaining issues**
- Need to talk about whether to delete the part of the spatial LDA notebook that does kmeans

### CODE REVIEW

**Comment by cliu72** [comment]
@ngreenwald wtf

File: templates_ark/example_neighborhood_analysis_script.ipynb [hunk 1]
```diff
@@ -1,18 +1,32 @@
 {
```
Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>Given that there are a lot of steps now for the notebook, having a summary here of what each step does would be helpful</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    all_fovs = list(all_data[settings.FOV_ID].drop_duplicates().values)
</pre><p><code>all_data[setting.FOV_ID].unique()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>See if you can use a smaller heading size for this text. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>What about adding a min_k as well? Did you ever look at <code>k</code> values below 5? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>It would be good to have an typical example of what this plot looks like in the notebook to give people an idea of what they're looking for</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>Same thing here, can you include an example output?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>Grande text size again</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (ngreenwald) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>Is the goal to replace all of this with Mantis once we have the single directory saving implemented? Or some of these plots should be kept even once we make that change?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

File: ark/analysis/spatial_analysis.py [hunk 2]
```diff
@@ -525,22 +521,28 @@ def create_neighborhood_matrix(all_data, dist_matrices_dict, included_fovs=None,
...
-        # add to neighbor counts + freqs for only the matching phenos between fov and whole dataset
+        # Add to neighbor counts + freqs for only the matching phenos between fov and whole dataset
         cell_neighbor_counts.loc[current_fov_neighborhood_data.index, fov_cluster_names] = counts
         cell_neighbor_freqs.loc[current_fov_neighborhood_data.index, fov_cluster_names] = freqs
 
-    # drop label column, as this interferes with the neighborhood clustering step
-    cell_neighbor_counts = cell_neighbor_counts.drop(columns=cell_label_col)
-    cell_neighbor_freqs = cell_neighbor_freqs.drop(columns=cell_label_col)
+    # Remove cells that have no neighbors within the distlim
+    cell_neighbor_counts = cell_neighbor_counts.loc[cell_neighbor_counts.drop(
```
Comment (ngreenwald) [inline ark/analysis/spatial_analysis.py:532]: This is a bit long for a one-liner. Since the `sum(axis=1) != 0` mask should be the same for both dfs, maybe you can pull that out into a separate initial line and use that to index both dfs?

File: ark/analysis/spatial_analysis.py [hunk 3]
```diff
@@ -593,17 +602,22 @@ def generate_cluster_matrix_results(all_data, neighbor_mat, cluster_num, exclude
...
+    neighbor_mat_data_all = neighbor_mat[neighbor_mat[fov_col].isin(included_fovs)]
+    neighbor_mat_data = neighbor_mat_data_all.drop([fov_col, label_col], axis=1)
 
     # generate cluster labels
     cluster_labels = spatial_analysis_utils.generate_cluster_labels(
         neighbor_mat_data, cluster_num)
 
-    all_data_clusters = all_data.copy()
+    # add labels to neighbor mat, add 1 to avoid cluster number 0
+    neighbor_mat_data_all[cluster_label_col] = [x+1 for x in cluster_labels]
```
Comment (ngreenwald) [inline ark/analysis/spatial_analysis.py:613]: Will you open an issue saying that `generate_cluster_labels` should default to 1-based cluster indices? Don't need to change it here if you don't want, but it would be a good default behavior to have eventually

File: ark/analysis/spatial_analysis_test.py [hunk 4]
```diff
@@ -332,16 +332,38 @@ def test_create_neighborhood_matrix():
...
+    # test if self_neighbor is False (default)
     counts, freqs = spatial_analysis.create_neighborhood_matrix(
         all_data_pos, dist_mat_pos, distlim=51
     )
 
-    # Test the counts values for both fovs
-    assert (counts.loc[:9, "Pheno2"] == 8).all()
-    assert (counts.loc[10:19, "Pheno3"] == 8).all()
+    # test the counts values
+    assert (counts[(counts[settings.FOV_ID] == "fov8") &
```
Comment (ngreenwald) [inline ark/analysis/spatial_analysis_test.py:341]: Were the tests not using two FOVs before? Why are we now indexing for specific FOVs, and specific segmentation labels within those FOVs?

File: ark/utils/spatial_analysis_utils.py [hunk 5]
```diff
@@ -435,39 +437,82 @@ def compute_neighbor_counts(current_fov_neighborhood_data, dist_matrix, distlim,
...
     # create the 'phenotype has cell?' matrix, excluding non cell-label rows
-    pheno_has_cell = pd.get_dummies(current_fov_neighborhood_data.iloc[:, 2]).to_numpy().T
+    pheno_has_cell_pd = pd.get_dummies(current_fov_neighborhood_data.loc[:, cluster_name_col])
+    pheno_names_from_tab = pheno_has_cell_pd.columns.values
+    pheno_has_cell = pheno_has_cell_pd.to_numpy().T
 
     # dot binarized 'is neighbor?' matrix with pheno_has_cell to get counts
     counts = pheno_has_cell.dot(cell_dist_mat_bin).T
+    counts_pd = pd.DataFrame(counts, columns=pheno_names_from_tab)
+    counts_pd = counts_pd.set_index(current_fov_neighborhood_data.index.copy())
```
Comment (ngreenwald) [inline ark/utils/spatial_analysis_utils.py:455]: Is this for code readability or is fixing a bug?

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looking great, some additional small comments here

File: templates_ark/example_neighborhood_analysis_script.ipynb (hunk 1 above)
Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: I'm not sure why it's showing up as huge in ReviewNB. It looks fine on github (https://github.com/angelolab/ark-analysis/blob/kmeans_neighborhood_nb_update/templates_ark/example_neighborhood_analysis_script.ipynb) and looks fine in my jupyter.

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Same as above, it looks fine on github (https://github.com/angelolab/ark-analysis/blob/kmeans_neighborhood_nb_update/templates_ark/example_neighborhood_analysis_script.ipynb) and looks fine in my jupyter.

File: ark/utils/spatial_analysis_utils.py (hunk 5 above)
Comment (cliu72) [reply ark/utils/spatial_analysis_utils.py:455]: Fixing a bug. Setting the columns is important - not all FOVs have the same cell types. This was the problem before that was leading to the wrong results. Setting the index is needed to merge the tables for all FOVs.

File: templates_ark/example_neighborhood_analysis_script.ipynb (hunk 1 above)
Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: If I understand it correctly, the mantis function (create_mantis_dir) just copies all the masks over, so the single directory saving will just take these masks and copy them over. This function actually creates the overlays.

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Yes, I have looked at k values below 5 when looking at specific regions of the lymph node (i.e. what are the neighborhoods only within the T cell zone?). But I can add a min_k anyways for flexibility.

File: ark/analysis/spatial_analysis_test.py (hunk 4 above)
Comment (cliu72) [reply ark/analysis/spatial_analysis_test.py:341]: I honestly don't know what was going on with the tests before. After fixing the bug in the code, those tests were just straight up wrong. I manually confirmed that this version is correct (literally just counted the values in the data table). Feel free to check my work.

File: ark/analysis/spatial_analysis.py (hunk 3 above)
Comment (cliu72) [reply ark/analysis/spatial_analysis.py:613]: I'll just change it.

File: templates_ark/example_neighborhood_analysis_script.ipynb (hunk 1 above)
Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Done

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Done

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Done

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: Done

File: ark/analysis/spatial_analysis.py (hunk 2 above)
Comment (cliu72) [reply ark/analysis/spatial_analysis.py]: Done

File: ark/analysis/spatial_analysis_test.py (hunk 4 above)
Comment (ngreenwald) [reply ark/analysis/spatial_analysis_test.py:341]: Yeah, sometimes ReviewNB does weird stuff. Once you've addressed all of the comments hit the re-request review button at the top, that way I know you're actually finished, easier for you than responding to each comment individually.

File: ark/utils/spatial_analysis_utils.py [hunk 6]
```diff
@@ -435,60 +437,118 @@ def compute_neighbor_counts(current_fov_neighborhood_data, dist_matrix, distlim,
...
+        if subsample is None:
+            cluster_score = sklearn.metrics.silhouette_score(neighbor_mat_data, cluster_labels,
+                                                             metric="euclidean")
+        else:
+            # Subsample each cluster
+            sub_dat = neighbor_mat_data.groupby("cluster").apply(
+                lambda x: x.sample(subsample, replace=len(x) < subsample)).reset_index(drop=True)
+            cluster_score = sklearn.metrics.silhouette_score(sub_dat.drop("cluster", axis=1),
+                                                             sub_dat["cluster"],
+                                                             metric="euclidean")
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply ark/utils/spatial_analysis_utils.py:543]: I think this also fixes a bug where the `'cluster'` column wasn't dropped from `neighbor_mat_data` when passing it to `silhouette_score` in the `subsample is None` case.

[the remark this answers is not in the document]
Comment (cliu72) [reply ark/utils/spatial_analysis_utils.py]: I don't think the original version of the code has that bug, since the `cluster` column is not dropped from `neighbor_mat_data` if `subsample is None`.

I made these suggested changes anyways.

**Review by alex-l-kong — CHANGES_REQUESTED** [review]
Made some additional comments, mostly regarding the notebook.

File: templates_ark/example_neighborhood_analysis_script.ipynb [hunk 7]
```diff
@@ -1,18 +1,34 @@
...
```
Comment (alex-l-kong) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>For each of these cells in section 4 we should have an example visualization.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (alex-l-kong) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>The notebook as it appears on GitHub has a blank cell after it, this should be removed.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (alex-l-kong) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>I think it's best to raise a <code>ValueError</code> here to make it consistent with similar validation checks. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (alex-l-kong) [inline templates_ark/example_neighborhood_analysis_script.ipynb:1]: <p>Is there any way we can adjust the colorbar width so it doesn't get too close to the heatmap?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/ark-analysis/pull/714/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

**Review by alex-l-kong — COMMENTED** [review]
Looks good, just a couple more minor comments on the notebook.

File: templates_ark/example_neighborhood_analysis_script.ipynb (hunk 7 above)
Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: This is an issue with all places where `visualize.draw_heatmap` is called. I noticed it when testing the cell clustering notebook. I believe it is also used in the spatial enrichment notebook. I think we can leave this for another pull request, so whoever makes the fix can confirm the heatmaps look okay in all of those notebooks.

Comment (cliu72) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: I also noticed just now that some of the other example notebooks do not have example output visualizations, as you have requested for this one. I think the pixel clustering and cell clustering notebooks, and spatial enrichment notebooks would also benefit from example outputs.

Comment (alex-l-kong) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: @cliu72 seems like the sample visualizations I had in the pixel and cell clustering notebooks got removed during some recent PRs, I'll ask @srivarra put those back in while he addresses the example dataset PR on his end.

Comment (alex-l-kong) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: > This is an issue with all places where `visualize.draw_heatmap` is called. I noticed it when testing the cell clustering notebook. I believe it is also used in the spatial enrichment notebook. I think we can leave this for another pull request, so whoever makes the fix can confirm the heatmaps look okay in all of those notebooks.

Yeah, the underlying `heatmap` function unfortunately doesn't offer great documentation in order to fix this. I think it'll require looking under the hood to see what internal properties are being set, I had to do that with a different `seaborn` function somewhere.

Comment (alex-l-kong) [reply templates_ark/example_neighborhood_analysis_script.ipynb:1]: @srivarra yes, the overlays for both notebooks as well as the weighted channel heatmaps for the cell clustering notebook.

**Comment by cliu72** [comment]
Addressed all comments. @ngreenwald @ackagel anything to add?

**Review by ngreenwald — APPROVED** [review]
Looks good!
